src/metabase/upload.clj

```diff
@@ -74,53 +75,38 @@
 ;;
 ;; </code></pre>
 
-(def ^:private type+parent-pairs
-  ;; listed in depth-first order
-  '([::boolean-or-int ::boolean]
-    [::boolean-or-int ::int]
-    [::auto-incrementing-int-pk ::int]
-    [::int ::float]
-    [::date ::datetime]
-    [::boolean ::varchar-255]
-    [::offset-datetime ::varchar-255]
-    [::datetime ::varchar-255]
-    [::float ::varchar-255]
-    [::varchar-255 ::text]))
-
-(defn ^:private column-type
-  "Returns the type of a column given the lowest common ancestor type of the values in the column."
-  [type]
-  (case type
-    ::boolean-or-int ::boolean
-    type))
-
-(def ^:private type->parents
-  (reduce
-   (fn [m [type parent]]
-     (update m type conj parent))
-   {}
-   type+parent-pairs))
+(def ^:private h
+  (-> (make-hierarchy)
+      (derive ::boolean-or-int ::boolean)
+      (derive ::boolean-or-int ::int)
+      (derive ::auto-incrementing-int-pk ::int)
+      (derive ::int ::float)
+      (derive ::date ::datetime)
+      (derive ::boolean ::varchar-255)
+      (derive ::offset-datetime ::varchar-255)
+      (derive ::datetime ::varchar-255)
+      (derive ::float ::varchar-255)
+      (derive ::varchar-255 ::text)))
+
+(def ^:private concrete-type?
+  "Nodes which can correspond to a concrete database column type."
+  (complement #{::boolean-or-int}))
 
 (def ^:private value-types
-  "All value types including the root type, ::text"
-  (conj (keys type->parents) ::text))
+  "All type tags which values can be inferred as. An ordered set from most to least specialized."
+  (ordered-hierarchy/sorted-tags h))
 
 (def ^:private column-types
-  "All column types"
-  (map column-type value-types))
-
-(defn- bfs-ancestors [type]
-  (loop [visit   (list type)
-         visited (ordered-set/ordered-set)]
-    (if (empty? visit)
-      visited
-      (let [parents (mapcat type->parents visit)]
-        (recur parents (into visited parents))))))
-
-(def ^:private type->bfs-ancestors
-  "A map from each type to an ordered set of its ancestors, in breadth-first order"
-  (into {} (for [type value-types]
-             [type (bfs-ancestors type)])))
+  "All type tags that correspond to concrete column types. An ordered set from most to least specialized."
+  (into #{} (filter concrete-type?) value-types))
+
+(defn ^:private column-type
+  "The most specific column type corresponding to the given value type."
+  [value-type]
+  ;; If we know nothing about the value type, treat it as an arbitrary string.
+  (cond (nil? value-type) ::text
+        (concrete-type? value-type) value-type
+        :else (recur (first (parents h value-type)))))
```

Comment:
Do you have a test for this recursion?

Reply:
It's well covered by the high level tests, but I don't mind adding a unit test.

Reply:
Really? what test will break if there's no recursion i.e.

```
:else (first (parents h value-type))
```

currently there's only one non-concrete type, so I can't see how this will break any high level tests

Reply:
I'm sure you know this, but explaining my reasoning - if a file has a column with only 0s or 1s (`::boolean-or-int`), we rely on this line (coerce to `::boolean`) to avoid a runtime error when we call `driver/upload-type->database-type` - in my comment I assumed we have at least one such.

Checking now, I see we only have a test that appends these values to an existing table. This gap in the testing is surprising to me. 

I'll write a unit test for this `column-type`.